[triton ci] Allow building with triton hash

PaliC opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131374
* __->__ #131371

In order to run inductor tests on Triton I am planning on using the infra we have already built on pytorch.  However, in order to do this we need to input a custom Triton pin as we need to build triton from source. This PR creates a mechanism to accomplish this.

Test: Inductor tests seem to build on https://github.com/pytorch/pytorch/pull/131374, but FAIL. Which is an indicator that this works :)

PaliC committed:
[triton ci] Allow building with triton hash

[ghstack-poisoned]

PaliC committed:
Update on "[triton ci] Allow building with triton hash"




[ghstack-poisoned]

PaliC committed:
Update on "[triton ci] Allow building with triton hash"




[ghstack-poisoned]

PaliC committed:
Update on "[triton ci] Allow building with triton hash"




[ghstack-poisoned]

huydhn commented on .ci/pytorch/build.sh:
```diff
@@ -276,6 +276,10 @@ else
   ( ! get_exit_code python setup.py clean bad_argument )
 
+  if [[ -n "$TRITON_VERSION" ]]; then
+    checkout_install_triton_at_commit "$TRITON_VERSION"
```
What is blocking us from running this inside Docker image build as the commit pin https://github.com/pytorch/pytorch/blob/main/.ci/docker/ci_commit_pins/triton.txt is already there?  IMO, only stuffs like vision needs to be built and installed at CI build time because they depend on PyTorch

PaliC replied:
Sorry this was a bit premature of a review, I'll put the context in the summary and request a another review.

malfet commented on .ci/pytorch/common_utils.sh:
```diff
@@ -208,6 +208,28 @@ function clone_pytorch_xla() {
+  fi
+  local commit
+  commit=$1
+  pushd ..
+  git clone --recurse-submodules https://github.com/triton-lang/triton.git
+  pushd triton
+
+  git checkout "${commit}"
+  # taken from triton README
+  pip install ninja cmake wheel; # build-time dependencies
+  pip install -e python
+  popd
+  popd
```
What's wrong with a single line
pip install --force-reinstall "git+https://github.com/trition-lang/trition@${commit}#subdirectory=python"

Also, why not reuse /extend install_trition_wheel.sh?

PaliC replied:
I didn't realize you could do that. I'll modify install_triton_wheel to do this.

PaliC committed:
Update on "[triton ci] Allow building with triton hash"


In order to run inductor tests on Triton I am planning on using the infra we have already built on pytorch.  However, in order to do this we need to input a custom Triton pin as we need to build triton from source. This PR creates a mechanism to accomplish this.

Test: Inductor tests seem to work on https://github.com/pytorch/pytorch/pull/131374

[ghstack-poisoned]
